README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# simulariatools

<!-- badges: start -->
<!-- [![R-CMD-check](https://github.com/Simularia/simulariatools/workflows/R-CMD-check/badge.svg)](https://github.com/Simularia/simulariatools/actions) -->
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.165117.svg)](https://doi.org/10.5281/zenodo.596741)
<!-- badges: end -->

`simulariatools` is an open source collection of functions and tools
useful to pre and post process data for air quality modelling and
assessment. The package is developed and maintained by the people at
[Simularia](https://www.simularia.it) and it is extensively used therein
for daily jobs.

If you use this package in your work, please consider citing it. Refer
to its [Zenodo DOI](https://doi.org/10.5281/zenodo.596741) to cite it.

## simulariatools Table of Contents

-   [Installation](#installation)
-   [Brief examples](#brief_examples)
-   [List of Functions](#list_of_functions)
-   [Contact](#contact)
-   [Contributors](#contributors)
-   [License](https://github.com/Simularia/simulariatools/blob/master/LICENSE.md)
-   [Changelog](https://github.com/Simularia/simulariatools/blob/master/NEWS)

## Installation

Installation of `simulariatools` from GitHub can be easily done using
the `devtools` package:

``` r
    library("devtools")
    install_github("Simularia/simulariatools")
```

The same commands can also be used to upgrade the package.

Note: in order to use `importADSOBIN()` to import *ADSO/BIN* data files,
a working installation of *Python 3* is required. For more information
about *R* and *Python* interoperability, please refer to
[`reticulate`](https://rstudio.github.io/reticulate/) documentation.

## Brief examples

### Contour plot

Firt, import air quality data from *NetCDF* or *ADSO/BIN* files with the
appropriate convenience function:

``` r
library(simulariatools)
mydata <- importRaster(file = "./test/conc_avg.nc",
                       k = 1000,
                       destaggering = TRUE,
                       variable = "nox",
                       verbose = TRUE)
#> Loading required namespace: ncdf4
#> 
#> Raster statistics -----------------------------------------------
#>        X (min, max, dx)  :  496000.000   519250.000      250.000
#>        Y (min, max, dy)  : 4943000.000  4955250.000      250.000
#>      nox (min, max, mean):    0.00e+00     2.71e+00     1.52e-01
#> -----------------------------------------------------------------
```

A quick contour plot with default configuration can be easily obtained
by running *contourPlot2()* without any argument:

``` r
contourPlot2(mydata)
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" /> The
plot is customisable by using *contourPlot2()* arguments and by piping
*ggplot2* instructions:

``` r
library(ggplot2)
contourPlot2(mydata, 
             domain = c(502000, 519000, 4943125, 4955125, 5, 5),
             levels = c(-Inf, 0.5, 1, 1.5, 2, Inf),
             legend = "NOx [ug/m3]") + 
  labs(x = NULL, y = NULL) +
  theme_minimal()
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />

Use `ggsave()` to save the last plot to file:

``` r
ggsave(filename = "~/path/to/myplot.png", width = 7, height = 6, dpi = 300)
```

If a *tile* plot is required:

``` r
library(ggplot2)
contourPlot2(mydata, 
             tile = TRUE,
             legend = "NOx [ug/m3]") + 
  labs(x = NULL, y = NULL) +
  theme_minimal()
```

<img src="man/figures/README-unnamed-chunk-7-1.png" width="100%" />

## List of functions

Available functions are listed below:

-   `contpourPlot2()`
-   `downloadBasemap()`
-   `importRaster()`.
-   `importADSOBIN()`.
-   `importSurferGrd()`.
-   `plotAvgRad()`.
-   `plotAvgTemp()`.
-   `plotStabilityClass()`.
-   `removeOutliers()`.
-   `rollingMax()`.
-   `stabilityClass()`.
-   `vectorField()`.
-   `contourPlot()` obsolete.
-   `createBaseMap()` obsolete.

## Contact

Contact person:

    Giuseppe Carlino
    Simularia s.r.l.
    g.carlino@simularia.it

## Contributors

Matteo Paolo Costa


NEWS
simulariatools 2.3.1
====================

- Trying to fix issues with fonts in contourPlot2().
- Updated readme file and fixed some minor issues in the documentation.

simulariatools 2.3.0
====================

- BREAKING CHANGE: in raster plots in `contourPlot2` (tile = TRUE) intervals are closed on the 
left and open on the right. In previous versions it was the opposite. In this way an interval
`0 - 1` does include values equal to 0 and does not include values equale to 1.

simulariatools 2.2.3
====================

- `magick` package is now suggested (not required). If it is not installed, it will not be possible
to read (and plot) background basemaps in `contourPlot2`.

- In `contourPlot2` do not add Inf as upper limit in levels in the case of delta maps, when the 
first level is negative. Note that this will probably change the output of some scripts.

- Vector field lengths are not normalised anymore and changed default scale factor. This modification
aims to facilitate comparison of vector fields from different source files.

- Updated readme and added a brief example.


simulariatools 2.2.2
====================

- Fixed labels of `contourPlot2` which occasionally were not properly formatted.

simulariatools 2.2.1
====================

- Minor update in the documentation.

simulariatools 2.2.0
====================

- Added boolean option `tile` to plot tiles instead of contours.

simulariatools 2.1.1
====================

* Fixed bug in `importRaster()`.

simulariatools 2.1.0
====================

* Breaking changes:
    - `vectorField()` has now a unique parameter for scaling vector length;
    - Length of the arrow head in `vectorField()` has been removed in exchange for a sensible default;
* Bug fixes.


simulariatools 2.0.0
====================

NEW FEATURES

* New `contourPlot2()` function:
* Based on ggplot2 v3.3.0 and `geom_contour_filled` function;
* Colour bands are not overlapped anymore therefore the maps are more accurate;
* Legend labels are more informatvie and smarter (you can also use +/- Inf for upper and lower bound);
* Legend type for contour lines (no filling) is now a line;
* Added possibility to specify column names to the data;
* Axis labels are rounded to 0 decimals (they are in metres);
* Backround image can be any format (jpeg, png, etc...);
* Almost entirly a substitute of `contourPlot()`.

* New `downloadBasemap()` function to download basemap from the Italian National Geoportal.

BUG FIXES AND MINOR IMPROVEMENTS

* `vectorField` has better defaults for vector lengths.
* Changed position of legend in `plotAvgRad()`.
* The package is now compliant with R cmd check (CRAN).
* Many other minor bugfixes and checks.

simulariatools 1.4.0
====================

* Introduced function `vectorField` to plot meteorological vector field.

simulariatools 1.3.1
=======================

* Function `importADSOBIN` is now compatible with `arinfopy` version 2.3.2.

simulariatools 1.3.0
=======================

* Function `contourPlot` accepts a **size** parameter for contour line thickness.
* Function `contourPlot` accepts a **cover** boolean parameter to specify if contours should be colour covered or not.
* Bugfix in `plotAvgRad`.

simulariatools 1.2.1
====================

* Function `contourPlot` accepts a **bare** boolean parameter to remove any graphical element of
the plot (axis, titles, legend, ...).

simulariatools 1.2.0
====================

* Function `contourPlot` accepts a **smoothness** parameter to improve the horizontal resolution of the dataset (smaller cells) by the given (integer) factor. The new raster object to be plotted is obtained by bilinear interpolation. 

simulariatools 1.1.0
====================

* New `importADSOBIN` function to read ADSO/BIN files.
* `plotStabilityClass` uses seasons instead of quarters.
* `plotAvgTemp` correctly manages timezones.
* Option `verbose of `importRaster` set to `FALSE` by default (it was `TRUE`).
* Option `fname` of `importRaster` renamed to `file` and set to `file.chooser` by default.
* Documentation improvement.


simulariatools 1.0.7
====================

* `plotStabilityClass`:
    - Correct order in color bars from 'A' to 'F'.
    
    
simulariatools 1.0.6
====================

* `contourPlot`:
    - Fixed bug when lowest level to plot is negative, as in delta maps;
    - Fixed bug when there is an overlay to plot. 


simulariatools 1.0.5
====================

* Borders in contour plots have width set to 0px by default.


simulariatools 1.0.4
====================

* Improved treatment of borders in `contourPlot`.


simulariatools 1.0.3
====================

* Fixed bug when extending domain to close contour lines.


simulariatools 1.0.2
====================

* Fixed bug in user defined colors for `contourPlot`.


simulariatools 1.0.1
====================

* New `colors` parameter to `contourPlot` for user defined color palette.


simulariatools 1.0
==================

* New `geom_hollow_polygon` and `stat_hollow_contour` to correctly plot hollows in filled contour plots.


simulariatools 0.6.5
====================

* Add verbose option in `importRaster`.
* Add option `kz` in `importRaster` to multiply `z` values.

simulariatools 0.6.4
====================

* Added transparency level in `contourPlot` input.


simulariatools 0.6.3
====================

* `contourPlot` is able to correctly plot negative values. Improved management of borders. Bugfixes.



simulariatools 0.6.2
====================

* Vertical shift of the labels in the contourPlot legend.


simulariatools 0.6.1
====================

* Fixed library dependence for `plotAvgTemp`.

* Added stMeteo dataset for testing.

simulariatools 0.6
====================

* New `contourPlot` function to plot contour levels diagrams of pollutant concentration.

* Code cleaning

simulariatools 0.5.3
====================

* bugfixes

simulariatools 0.5.2
====================

* bugfixes

simulariatools 0.5.1
====================


R/contourPlot.R
#' Contour plot of pollutant concentration
#'
#' \code{contourPlot} plots a contour map of pollutants.
#'
#' This is a convenience function to plot contour levels of a pollutant matrix
#' with \code{ggplot2}.
#'
#' @param data A dataframe containing data to be plotted in the form of X, Y and
#'   Z (levels).
#' @param domain An array with min X, max X, min Y, max Y, number of ticks on X
#'   axis, number of ticks on Y axis (optional).
#' @param background String containing the path to the png file to be plotted as
#'   a basemap (optional).
#' @param underlayer Array of strings containing layers to be plotted between
#'   basemap and contour plot (optional).
#' @param overlayer Array of strings containing layers to be plotted on top of
#'   the contour plot (optional).
#' @param legend (string) Legend title (optional).
#' @param levels Array of levels for contour plot. If not set, automatic levels
#'   are plotted.
#' @param transparency float (between 0 and 1, default=0.66). Transparency level
#'   of the contour plot.
#' @param smoothness integer factor to improve the horizontal resolution 
#'   (smaller cells) by bilinear interpolation.
#' @param colors Color palette for contour plot
#' @param bare Boolean (default FALSE) parameter to completely remove axis,
#'  legend, titles and any other graphical element from the plot.
#' @param size float with the thickness of the contour line.
#' @param cover boolean (default TRUE) to specify whether the contour plot 
#'   should be filled or not.
#'
#' @return A \code{ggplot2} plot.
#' 
#' @examples
#' \dontrun{
#' # Import variable CONCAN from inpufile, convert km to m (k = 1000):
#' data <- importRaster(paste0(dir, inputfile), 
#'                      k = 1000, 
#'                      variable = "CONCAN")
#'
#' # Simple contour plot
#' contourPlot(data)
#'
#' # Specifiy (sub)domain to be plotted; background image; legend title and 
#' pollutant levels.
#' contourPlot(data, 
#'             domain(500000, 510000, 6000000, 6010000, 7, 7), 
#'             background = "img/background.png", 
#'             legend = "no2 [ug/m3]", 
#'             levels = c(10, 20, 30, 40))
#'
#' # Add underlayer (same for overlayer)
#' library(ggplot2)
#' library(maptools)
#' perimetro <- readShapeLines("path_to/perimetro.shp")
#' perimetro <- fortify(perimetro)
#' strada <- readShapeLines("path_to/strada.shp")
#' strada <- fortify(strada)
#' myUnderlayer <- vector(mode = "list", length = 2)
#' myUnderlayer[[1]] <- geom_polygon(data = perimetro, 
#'                                   aes(long, lat, group = group), 
#'                                   colour = "black", 
#'                                   fill = NA, 
#'                                   size = 0.1, 
#'                                   alpha = 0.5)
#' myUnderlayer[[2]] <- geom_path(data = strada, 
#'                                aes(long, lat, group = group), 
#'                                colour = "grey", 
#'                                size = 0.1, 
#'                                alpha = 0.5)
#' contourPlot(data = test, 
#'             background = "path_to/basemap.png", 
#'             underlayer = myUnderlayer)
#'
#' # Change default colour palette
#' contourPlot(data = test, 
#'             colors = RColorBrewer::brewer.pal(3, name = "PiYG"))
#' }
#'
#' @import ggplot2
#' @importFrom grDevices colorRampPalette
#' @importFrom raster rasterFromXYZ disaggregate xmin xmax ymin ymax values
#'                    crop merge extent extend rasterToPoints
#' 
#' @export
#' 
contourPlot <- function(data,
                        domain = NULL,
                        background = NULL, 
                        underlayer = NULL,
                        overlayer = NULL,
                        legend = NULL, 
                        levels = NULL,
                        size = 0.,
                        cover = TRUE,
                        transparency = 0.66,
                        smoothness = 1.,
                        colors = NULL,
                        bare = FALSE) {
  
  .Deprecated("contourPlot2")

  # Local binding for variables
  x <- y <- z <- NULL
  
  # Convert input to raster
  tt <- raster::rasterFromXYZ(data)
  
    # Resample raster
    tt <- raster::disaggregate(tt, fact = smoothness, method = 'bilinear')
    
    # Define plot domain
        if (missing(domain)) {
            xmin <- raster::xmin(tt)    # x coordinates minimum
            xmax <- raster::xmax(tt)    # x coordinates max
            ymin <- raster::ymin(tt)    # y coordinates min
            ymax <- raster::ymax(tt)    # y coordinates max
            nx <- 5                 # number of ticks along x axis
            ny <- 5                 # number of ticks along y axis
        } else {
            xmin <- domain[1]       # x coordinates minimum
            xmax <- domain[2]       # x coordinates max
            ymin <- domain[3]       # y coordinates min
            ymax <- domain[4]       # y coordinates max
            nx <- domain[5]         # number of ticks along x axis
            ny <- domain[6]         # number of ticks along y axis
        }
    
    # Automatic scales
    if (missing(levels)) {
        nlevels <- 7
        levels <- pretty(range(raster::values(tt), na.rm = TRUE),
                         n = nlevels, min.n = 4)
    }
    lab_levels <- levels

    # Extend data domain to be plotted
    for (i in (1:1)) {

        for (idx in seq(1, 1)) {
            # Extend top and boottom rows
            ttx1 <- raster::crop(tt, raster::extent(tt, 1, 1, 1,
                                                    raster::ncol(tt)))
            raster::ymin(ttx1) <- raster::ymax(tt)
            raster::ymax(ttx1) <- raster::ymax(tt) + raster::res(tt)[2]
            ttxN <- raster::crop(tt, raster::extent(tt, raster::nrow(tt),
                                                    raster::nrow(tt), 1, 
                                                    raster::ncol(tt)))
            raster::ymax(ttxN) <- raster::ymin(tt)
            raster::ymin(ttxN) <- raster::ymin(tt) - raster::res(tt)[2]
            ttE <- raster::merge(tt, ttx1)
            ttE <- raster::merge(ttE, ttxN)

            # Extend left and right columns
            tty1 <- raster::crop(tt, raster::extent(tt, 1,
                                                    raster::nrow(tt), 1, 1))
            raster::xmin(tty1) <- raster::xmin(tt) - raster::res(tt)[1]
            raster::xmax(tty1) <- raster::xmin(tt)
            ttyN <- raster::crop(tt, raster::extent(tt, 1, 
                                                    raster::nrow(tt),
                                                    raster::ncol(tt),
                                                    raster::ncol(tt)))
            raster::xmin(ttyN) <- raster::xmax(tt)
            raster::xmax(ttyN) <- raster::xmax(tt) + raster::res(tt)[1]
            ttE <- raster::merge(ttE, tty1)
            ttE <- raster::merge(ttE, ttyN)
            tt <- ttE
        }
        et <- raster::extent(raster::xmin(ttE) - 1 * raster::res(tt)[1],
                             raster::xmax(ttE) + 1 * raster::res(tt)[1],
                             raster::ymin(ttE) - 1 * raster::res(tt)[2],
                             raster::ymax(ttE) + 1 * raster::res(tt)[2])
        mv <- min(raster::values(tt), na.rm = TRUE)
        j <- 1
        while (lab_levels[j] < mv) {
            j <- j + 1
        }
        if (j != 1) {
            ev <- lab_levels[j - 1]
        } else {
            ev <- lab_levels[1]
        }
        # ev <- ev - ev/10.
        if (lab_levels[1] < 0.) {
            ev <- lab_levels[1] - 1.
        }

        ttE <- raster::extend(ttE, et, value = ev)
    }
    
    # convert raster to dataframe 
    ttP <- raster::rasterToPoints(ttE)
    ttDF <- data.frame(ttP)
    colnames(ttDF) <- c("x", "y", "z")
    
    # boundaries of extended domain
    xminE <- raster::xmin(ttE)
    yminE <- raster::ymin(ttE)
    xmaxE <- raster::xmax(ttE)
    ymaxE <- raster::ymax(ttE)
    
    # color palette (omit first color)
    if (is.null(colors)) {
        myPalette <- grDevices::colorRampPalette(
            rev(RColorBrewer::brewer.pal(11, name = "Spectral")))
        myColors <- myPalette(length(levels) + 1)[-c(1, 1)]
    } else {
        myPalette <- grDevices::colorRampPalette(colors, alpha = TRUE)
        myColors <- myPalette(length(levels))
    }
    
    # Legend
    if (is.null(legend)) {
        legend <- ""
    }
    # prettify legend title
    # TODO: svincolarsi da openair per l'etichetta
    if (requireNamespace("openair", quietly = TRUE)) {
        lgndname <- openair::quickText(legend, auto.text = TRUE)
    } else {
        lgndname <- legend
    }

    # Background image
    if (missing(background)) {
        img <- matrix(data = NA, nrow = 10, ncol = 10)
        gimg <- grid::rasterGrob(img, interpolate = TRUE)
    } else {
        img <- png::readPNG(background)
        gimg <- grid::rasterGrob(img, interpolate = TRUE)
    }
    
    # Underlayer
    if (missing(underlayer)) {
        underlayer <- geom_blank()
    }
    
    # Overlayer
    if (missing(overlayer)) {
        overlayer <- geom_blank()
    }

    # If requested, wipe all but main plot 
    if (isTRUE(bare)) {
        baretheme <- theme(axis.line=element_blank(),
                           axis.text.x=element_blank(),
                           axis.text.y=element_blank(),
                           axis.ticks=element_blank(),
                           axis.title.x=element_blank(),
                           axis.title.y=element_blank(),
                           legend.position="none",
                           panel.background=element_blank(),
                           panel.border=element_blank(),
                           panel.grid.major=element_blank(),
                           panel.grid.minor=element_blank(),
                           plot.background=element_blank())
    } else {
        baretheme <- theme()
    }

    # Contour plot
    v <- ggplot(ttDF) +
        annotation_custom(gimg, -Inf, Inf, -Inf, Inf)  +
        underlayer + 
        stat_hollow_contour(
            data = ttDF,
            aes(x = x, y = y, z = z, fill = factor(..level..)),
            geom = "hollow_polygon",
            size = size,
            breaks = levels,
            alpha = transparency,
            cover = cover,
            na.rm = TRUE) +
        scale_fill_manual(lgndname,
                          guide = guide_legend(reverse = TRUE, label.vjust = 0),
                          breaks = levels,
                          labels = lab_levels,
                          values = myColors) +
        scale_x_continuous(name = "x [m]",
                           limits = c(xminE, xmaxE),
                           breaks = seq(xmin, xmax, length.out = nx),
                           expand = c(0, 0)) +
        scale_y_continuous(name = "y [m]",
                           limits = c(yminE, ymaxE),
                           breaks = seq(ymin, ymax, length.out = ny),
                           expand = c(0, 0)) +
        theme_bw(base_size = 10, base_family = "Arial") +
        coord_fixed(ratio = 1, xlim = c(xmin, xmax), ylim = c(ymin, ymax)) +
        overlayer +
        baretheme
  
    return(v)
}